pubsub/tests/unit/pubsub_v1/subscriber/test_leaser.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2017, Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import threading
+
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.subscriber import _histogram
+from google.cloud.pubsub_v1.subscriber import subscriber
+from google.cloud.pubsub_v1.subscriber._protocol import leaser
+from google.cloud.pubsub_v1.subscriber._protocol import requests
+
+import mock
+import pytest
+
+
+def test_add_and_remove():
+    leaser_ = leaser.Leaser(mock.sentinel.subscriber)
+
+    leaser_.add([
+        requests.LeaseRequest(ack_id='ack1', byte_size=50)])
+    leaser_.add([
+        requests.LeaseRequest(ack_id='ack2', byte_size=25)])
+
+    assert leaser_.message_count == 2
+    assert set(leaser_.ack_ids) == set(['ack1', 'ack2'])
+    assert leaser_.bytes == 75
+
+    leaser_.remove([
+        requests.DropRequest(ack_id='ack1', byte_size=50)])
+
+    assert leaser_.message_count == 1
+    assert set(leaser_.ack_ids) == set(['ack2'])
+    assert leaser_.bytes == 25
+
+
+def test_add_already_managed(caplog):
+    caplog.set_level(logging.DEBUG)
+
+    leaser_ = leaser.Leaser(mock.sentinel.subscriber)
+
+    leaser_.add([
+        requests.LeaseRequest(ack_id='ack1', byte_size=50)])
+    leaser_.add([
+        requests.LeaseRequest(ack_id='ack1', byte_size=50)])
+
+    assert 'already lease managed' in caplog.text
+
+
+def test_remove_not_managed(caplog):
+    caplog.set_level(logging.DEBUG)
+
+    leaser_ = leaser.Leaser(mock.sentinel.subscriber)
+
+    leaser_.remove([
+        requests.DropRequest(ack_id='ack1', byte_size=50)])
+
+    assert 'not managed' in caplog.text
+
+
+def test_remove_negative_bytes(caplog):
+    caplog.set_level(logging.DEBUG)
+
+    leaser_ = leaser.Leaser(mock.sentinel.subscriber)
+
+    leaser_.add([
+        requests.LeaseRequest(ack_id='ack1', byte_size=50)])
```

Comment:
More general questions:
- Are messages guaranteed to have unique ack ids
- Are messages immutable?

the first is more a general question. the second is about understanding how we end up in negative byte scenarios.